src/UsgsAstroSarSensorModel.cpp

```
@@ -136,3 +141,209 @@ string UsgsAstroSarSensorModel::constructStateFromIsd(
+                            + lookVector[2]*spacecraftVelocity[2])/(slantRange * WAVELENGTH);
+     return dopplerShift;
+   };
+
+  // Lower-bound for doppler-shift
+  double startShift = dopplerShiftFrequency(START_TIME);
+
+  // Upper-bound for doppler-shift
+  double stopShift = dopplerShiftFrequency(STOP_TIME);
+
+  // Make sure we bound root (dopplerShift = 0.0)
+  if (((startShift < 0.0) && (stopShift < 0.0)) || ((startShift > 0.0) && (stopShift > 0.0))) {
+    return -1;
+  }
+
+  // Do root-finding for "dopplerShift"
+  double tolerance = (STOP_TIME - START_TIME) / LINES / 20.0;
+  double maxIterations = 30;
+
+  // Order the bounds
+  double fl, fh, xl, xh;
+  if (startShift < stopShift) {
+    xl = START_TIME;
+    xh = STOP_TIME;
+  }
+  else {
+    xl = STOP_TIME;
+    xh = START_TIME;
+  }
+
+  return secantRoot(dopplerShiftFrequency, xl, xh, tolerance, 30);
+}
+
+
+double secantRoot(std::function<double(double)> lambda, double xl, double xh, 
+                  double tolerance, double maxIters) {
+  double time;
+  int j = 0;
+    int found = -1;
+    double shift;
+    double fl = lambda(xl);
+    double fh = lambda(xh);
+    while (found==-1 && j < maxIters) {
+      // Use the secant method to guess the next time
+      time = xl + (xh - xl) * fl / (fl - fh);
+      
+      // Compute a guessed Doppler shift.
+      shift = lambda(time);
+      
+      // Update the bounds
+      double deltaTime;
+      if (shift < 0.0) {
+        deltaTime = xl - time;
+        xl = time;
+        fl = shift;
+      }
+      else {
+        deltaTime = xh - time;
+        xh = time;
+        fh = shift;
+      }
+
+      // See if we are done
+      if ((fabs(deltaTime) <= tolerance) || (shift == 0.0)) {
+        time = time;
+        found = 0;
+      }
+      j++;
+    }
+    if (found != -1) {
+      return time; 
+    }
+    else {
+      return -1;
```

This should raise an exception.